lib/tree.rb

```diff
@@ -16,40 +16,129 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # def add(key, value)
+  #   if @root.nil?
+  #     @root = TreeNode.new(key, value)
+  #   else 
+  #     current = @root
+
+  #     while true
+  #       if key < current.key
+  #         if !current.left.nil?
+  #           current = current.left
+  #         else
+  #           current.left = TreeNode.new(key, value)
+  #           return
+  #         end
+  #       else
+  #         if !current.right.nil?
+  #           current = current.right
+  #         else
+  #           current.right = TreeNode.new(key, value)
+  #           return
+  #         end
+  #       end
+  #     end
+  #   end
+  # end
+
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+
+    return current_node
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(logn)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    current = @root
+
+    return nil if current.nil?
+
+    until current.key == key
+      if current.key > key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
```

Comment:
What about if the key is not in the tree?  What about when `current` is `nil`?

```suggestion
    until current.key == key
      if current.key > key
        current = current.left
      else
        current = current.right
      end
      return nil if current.nil?
    end
```